bchehraz/react-challenge-project-jan-2022
# Challenge 8: Add "edit" and "delete" order functionality

## Changes
_List Changes Introduced by this PR_
1. (from challenge 1) Fixes order form input by changing `event.value` to `event.target.value`
2. (from challenge 4) Fixes authReducer action payload to properly login
3. (from challenge 5) Wires in the logout button in the `nav` component.
4. (from challenge 6) Wires in `PrivateRoute` route guard component
5. Renames `order-form` to `order-form-page` and pulls out order-form into its own component. Additionally updates `order-form-page.test.js` description to match new component name.
6. Adds "delete" functionality to `ordersList.js`
7. Adds "edit" functionality to `order-form` which acts as an `edit` form when an order is passed in via `props`.
8. Wires in "edit" functionality back into the `ordersList.js` component.

## Purpose
_Describe the problem or feature in addition to a link to the issues._

- Link to issue: https://github.com/Shift3/react-challenge-project-jan-2022/issues/8
- Before this PR, the Edit and Delete buttons did nothing in the order list. The purpose was to wire in this functionality, but also keep the code as DRY as possible.

## Approach
_How does this change address the problem?_

- [Edit Feature] To keep the code as dry as possible, I figured it would be simple enough to use the `order-form` component as both a create and edit form. To accomplish this, I just pass an existing order via `props` and determine if it's an edit by checking for an existing `order._id` prop. Now I can display this form below an order when the "Edit" button is pushed on the orders list. However, since the `order-form` acted as more of a page (since it had a `Template` wrapper), I instead created an `order-form-page` component which referenced the `order-form`, wrapped with a Template. 
- [Delete Feature] This just entailed creating a function that called the `api/delete-order` endpoint and wire it into the "Delete" button in the order list by passing in the `_id` of the order.

## Pre-Testing TODOs
_What needs to be done before testing_
- [ ] Before testing, create several new orders on the `/order` page.

## Testing Steps
_How do the users test this change?_

1. Navigate to `/view-orders` and see your created orders. Push the "Delete" button to delete one of the orders to test the delete feature.
2. On the last remaining order, click the "Edit" button and make a modification to the order by selecting a new menu item and changing the quantity. Then push "Update" to finalize the changes. Now the order should reflect the changes you requested. Also notice the Edit form can only be opened for one order at a time.

Closes https://github.com/Shift3/react-challenge-project-jan-2022/issues/8
